### PR TITLE
Update CheckConsistency in DefaultFileSystemMaster for Metadata Sync

### PR DESCRIPTION
**Pull request by tahakashaf**
Added UFS Fingerprint check to decide a file in sync or not.
https://alluxio.atlassian.net/browse/ALLUXIO-3363

### CODE REVIEW

**Comment by tahakashaf** [comment]
alluxio-bot, check this please

**Comment by tahakashaf** [comment]
Jenkins, test this please.

File: core/server/master/src/main/java/alluxio/master/file/DefaultFileSystemMaster.java [hunk 1]
```diff
@@ -1197,7 +1197,8 @@ private boolean checkConsistencyInternal(LockedInodePath inodePath) throws Inval
         return ufs.isDirectory(ufsPath);
       } else {
         return ufs.isFile(ufsPath)
-            && ufs.getFileStatus(ufsPath).getContentLength() == inode.asFile().getLength();
+            && ufs.getFileStatus(ufsPath).getContentLength() == inode.asFile().getLength()
+            && (!ufs.getFingerprint(ufsPath).equals(Constants.INVALID_UFS_FINGERPRINT));
```
Comment (yuzhu) [inline core/server/master/src/main/java/alluxio/master/file/DefaultFileSystemMaster.java]: we could actually check if the fingerprint matches between the ufs and the inode, using `inode.getUfsFingerprint()`.

Comment (madanadit) [inline core/server/master/src/main/java/alluxio/master/file/DefaultFileSystemMaster.java]: we should compare `ufs.getFingerprint(ufsPath)` with `inode.asFile().getUfsFingerprint` if both are valid

Comment (madanadit) [reply core/server/master/src/main/java/alluxio/master/file/DefaultFileSystemMaster.java]: actually, this can be made more efficient by combining the ufs calls into ufs.getStatus:

ufsIsDirectory = ufsStatus.isDirectory()
ufsIfFile = ufsStatus.isFile()
ufsFingerprint = Fingerprint.create(ufs.getUnderFSType(), ufsStatus).serialize();

Comment (tahakashaf) [reply core/server/master/src/main/java/alluxio/master/file/DefaultFileSystemMaster.java]: @madanadit I have tried to use the ufsStatus in the below way 
`try (CloseableResource<UnderFileSystem> ufsResource = resolution.acquireUfsResource()) {
      UnderFileSystem ufs = ufsResource.get();
      String ufsPath = resolution.getUri().getPath();
      UfsStatus ufsStatus = ufs.getStatus(ufsPath);
      if (ufs == null) {
        return true;
      }
      if (!inode.isPersisted()) {
        return !ufs.exists(ufsPath);
      }
      // TODO(calvin): Evaluate which other metadata fields should be validated.
      if (inode.isDirectory()) {
        return ufsStatus.isDirectory();
      } else {
        String ufsFingerprint = Fingerprint.create(ufs.getUnderFSType(), ufsStatus).serialize();
        return ufsStatus.isFile()
            && ufs.getFileStatus(ufsPath).getContentLength() == inode.asFile().getLength()
            && (ufsFingerprint.equals(inode.asFile().getUfsFingerprint()));
      }`

Comment (tahakashaf) [reply core/server/master/src/main/java/alluxio/master/file/DefaultFileSystemMaster.java]: `try (CloseableResource<UnderFileSystem> ufsResource = resolution.acquireUfsResource()) {
      UnderFileSystem ufs = ufsResource.get();
      String ufsPath = resolution.getUri().getPath();
      UfsStatus ufsStatus = ufs.getStatus(ufsPath);
      if (ufs == null) {
        return true;
      }
      if (!inode.isPersisted()) {
        return !ufs.exists(ufsPath);
      }
      // TODO(calvin): Evaluate which other metadata fields should be validated.
      if (inode.isDirectory()) {
        return ufsStatus.isDirectory();
      } else {
        String ufsFingerprint = Fingerprint.create(ufs.getUnderFSType(), ufsStatus).serialize();
        return ufsStatus.isFile()
            && ufs.getFileStatus(ufsPath).getContentLength() == inode.asFile().getLength()
            && (ufsFingerprint.equals(inode.asFile().getUfsFingerprint()));
      }`

Comment (tahakashaf) [reply core/server/master/src/main/java/alluxio/master/file/DefaultFileSystemMaster.java]: > try (CloseableResource<UnderFileSystem> ufsResource = resolution.acquireUfsResource()) {
      UnderFileSystem ufs = ufsResource.get();
      String ufsPath = resolution.getUri().getPath();
      UfsStatus ufsStatus = ufs.getStatus(ufsPath);
      if (ufs == null) {
        return true;
      }
      if (!inode.isPersisted()) {
        return !ufs.exists(ufsPath);
      }
      // TODO(calvin): Evaluate which other metadata fields should be validated.
      if (inode.isDirectory()) {
        return ufsStatus.isDirectory();
      } else {
        String ufsFingerprint = Fingerprint.create(ufs.getUnderFSType(), ufsStatus).serialize();
        return ufsStatus.isFile()
            && ufs.getFileStatus(ufsPath).getContentLength() == inode.asFile().getLength()
            && (ufsFingerprint.equals(inode.asFile().getUfsFingerprint()));
      }

Comment (tahakashaf) [reply core/server/master/src/main/java/alluxio/master/file/DefaultFileSystemMaster.java]: Thanks @yuzhu

Comment (madanadit) [reply core/server/master/src/main/java/alluxio/master/file/DefaultFileSystemMaster.java]: @tahakashaf can you please update the PR. you can move the getStatus call to after the null check for ufs and catch FileNotFoundException from ufs.getStatus instead of the ufs.exists call

File: core/server/master/src/main/java/alluxio/master/file/DefaultFileSystemMaster.java [hunk 2]
```diff
@@ -2391,24 +2389,16 @@ public UfsInfo getUfsInfo(long mountId) {
...
-   * @throws FileAlreadyCompletedException if the file is already completed
-   * @throws FileDoesNotExistException if there is no UFS path
-   * @throws InvalidFileSizeException if invalid file size is encountered
-   * @throws InvalidPathException if invalid path is encountered
    */
   private void loadMetadataInternal(RpcContext rpcContext, LockedInodePath inodePath,
       LoadMetadataContext context)
-      throws AccessControlException, BlockInfoException, FileAlreadyCompletedException,
-      FileDoesNotExistException, InvalidFileSizeException, InvalidPathException, IOException {
+      throws InvalidPathException, FileDoesNotExistException, BlockInfoException,
```
Comment (madanadit) [inline core/server/master/src/main/java/alluxio/master/file/DefaultFileSystemMaster.java]: nit: I prefer the old alphabetically sorted order. please revert this change

File: core/server/master/src/main/java/alluxio/master/file/DefaultFileSystemMaster.java [hunk 3]
```diff
@@ -2391,24 +2389,16 @@ public UfsInfo getUfsInfo(long mountId) {
   /**
    * Loads metadata for the object identified by the given path from UFS into Alluxio.
    *
-   * This operation requires users to have WRITE permission on the path
```
Comment (madanadit) [inline core/server/master/src/main/java/alluxio/master/file/DefaultFileSystemMaster.java]: please revert this change

File: core/server/master/src/main/java/alluxio/master/file/DefaultFileSystemMaster.java [hunk 4]
```diff
@@ -435,15 +435,8 @@ public DefaultFileSystemMaster(BlockMaster blockMaster, CoreMasterContext master
...
-    mJournaledComponents = new ArrayList<Journaled>() {
-      {
-        add(mInodeTree);
-        add(mDirectoryIdGenerator);
-        add(mMountTable);
-        add(mUfsManager);
-        add(mSyncManager);
-      }
-    };
+    mJournaledComponents =
```
Comment (madanadit) [inline core/server/master/src/main/java/alluxio/master/file/DefaultFileSystemMaster.java]: nit:  please revert changes unrelated to checkConsistency in this PR

File: core/server/master/src/main/java/alluxio/master/file/DefaultFileSystemMaster.java [hunk 5]
```diff
@@ -1195,15 +1188,20 @@ private boolean checkConsistencyInternal(LockedInodePath inodePath) throws Inval
       if (ufs == null) {
         return true;
       }
-      if (!inode.isPersisted()) {
-        return !ufs.exists(ufsPath);
+      UfsStatus ufsStatus;
+      try {
+        ufsStatus = ufs.getStatus(ufsPath);
+      } catch (FileNotFoundException e) {
+        throw new RuntimeException(e);
```
Comment (madanadit) [inline core/server/master/src/main/java/alluxio/master/file/DefaultFileSystemMaster.java]: return false here and add back `inode.isPersisted` instead of throwing an exception

File: core/server/master/src/main/java/alluxio/master/file/DefaultFileSystemMaster.java [hunk 6]
```diff
@@ -1195,15 +1188,20 @@ private boolean checkConsistencyInternal(LockedInodePath inodePath) throws Inval
...
       // TODO(calvin): Evaluate which other metadata fields should be validated.
       if (inode.isDirectory()) {
-        return ufs.isDirectory(ufsPath);
+        return ufsStatus.isDirectory();
       } else {
-        return ufs.isFile(ufsPath)
-            && ufs.getFileStatus(ufsPath).getContentLength() == inode.asFile().getLength();
+        String ufsFingerprint = Fingerprint.create(ufs.getUnderFSType(), ufsStatus).serialize();
+        return ufsStatus.isFile()
+            && ufs.getFileStatus(ufsPath).getContentLength() == inode.asFile().getLength()
```
Comment (madanadit) [inline core/server/master/src/main/java/alluxio/master/file/DefaultFileSystemMaster.java]: `ufs.getFileStatus` is not required as you already have `ufsStatus`

File: core/server/master/src/main/java/alluxio/master/file/DefaultFileSystemMaster.java [hunk 7]
```diff
@@ -1193,15 +1193,22 @@ private boolean checkConsistencyInternal(LockedInodePath inodePath) throws Inval
       if (ufs == null) {
         return true;
       }
+      UfsStatus ufsStatus;
+      try {
+        ufsStatus = ufs.getStatus(ufsPath);
+      } catch (FileNotFoundException e) {
+        return false;
```
Comment (madanadit) [inline core/server/master/src/main/java/alluxio/master/file/DefaultFileSystemMaster.java]: this should be `return !inode.isPersisted()` if its not in the ufs. you could return early after line 1189, with if (!inode.isPersisted) return true

**Comment by madanadit** [comment]
Thanks @tahakashaf. I left one more comment.

**Comment by apc999** [comment]
@madanadit @yuzhu do you have more comments? Are we targeting to merge this pull request to 2.0?

**Comment by madanadit** [comment]
jenkins, test this please.

**Comment by madanadit** [comment]
@apc999 we should target 2.0. will merge as soon as test passes. @yuzhu let us know if you have more comments

**Comment by madanadit** [comment]
the test fails because local ufs does not through FileNotFoundException on getStatus when the path does not exist. fixing in a separate PR and will merge this after

**Comment by madanadit** [comment]
jenkins, test this please.

**Comment by madanadit** [comment]
Closed in favor of: https://github.com/Alluxio/alluxio/pull/8588

**Comment by madanadit** [comment]
Thanks @tahakashaf for your contribution. Opened a replacement PR here: https://github.com/Alluxio/alluxio/pull/9135
